src/data_source/sql.rs

```diff
@@ -0,0 +1,987 @@
+///
+/// ```
+/// # use hotshot_query_service::data_source::sql::{include_migrations, Migration};
+/// let migrations: Vec<Migration> = include_migrations!("$CARGO_MANIFEST_DIR/migrations").collect();
+/// assert_eq!(migrations[0].version(), 10);
+/// assert_eq!(migrations[0].name(), "init_schema");
+/// ```
+#[macro_export]
+macro_rules! include_migrations {
+    ($dir:tt) => {
+        $crate::data_source::sql::include_dir!($dir)
+            .files()
+            .map(|file| {
+                let path = file.path();
+                let name = path
+                    .file_name()
+                    .and_then(std::ffi::OsStr::to_str)
+                    .unwrap_or_else(|| {
+                        panic!(
+                            "migration file {} must have a non-empty UTF-8 name",
+                            path.display()
+                        )
+                    });
+                let sql = file
+                    .contents_utf8()
+                    .unwrap_or_else(|| panic!("migration file {name} must use UTF-8 encoding"));
+                Migration::unapplied(name, sql).expect("invalid migration")
+            })
+    };
+}
+
+/// The migrations requied to build the default schema for this version of [`SqlDataSource`].
+pub fn default_migrations() -> Vec<Migration> {
+    let mut migrations = include_migrations!("$CARGO_MANIFEST_DIR/migrations").collect::<Vec<_>>();
+    validate_migrations(&mut migrations, true).expect("default migrations are invalid");
+    migrations
+}
+
+/// Validate and preprocess a sequence of migrations.
+///
+/// * Ensure all migrations have distinct versions
+/// * Ensure migrations are sorted by increasing version
+/// * If validating [`default_migrations`] (i.e. `default` is `true`), additionally check that all
+///   migration versions are multiples of 10, so that custom migrations can be inserted in between.
+fn validate_migrations(migrations: &mut [Migration], default: bool) -> Result<(), Error> {
+    migrations.sort_by_key(|m| m.version());
+
+    // Check version uniqueness.
+    for (prev, next) in migrations.iter().zip(migrations.iter().skip(1)) {
+        if next <= prev {
+            return Err(Error::msg(format!(
+                "migration versions are not strictly increasing ({prev}->{next})"
+            )));
+        }
+    }
+
+    if default {
+        // Check version spacing.
+        for m in migrations {
+            if m.version() == 0 || m.version() % 10 != 0 {
+                return Err(Error::msg(
+                    "default migration versions are not positive multiples of 10",
+                ));
+            }
+        }
+    }
+
+    Ok(())
+}
+
+/// Merge two migration sequences.
+///
+/// Migrations in `right` replace migrations in `left` with the same version. Each of `left` and
+/// `right` is assumed to be the output of [`validate_migrations`]; that is, each is sorted by
+/// version and contains no duplicate versions.
+fn merge_migrations(
```

Comment:
This seems to be used only once to merge custom migrations into default migrations and is not exported. So I wonder if the code could be inlined instead or if we expect it to be useful elsewhere.

Reply:
I don't know if it will be useful anywhere, but I think it still makes sense as a separate function. It's a self-contained bit of logic and if I inlined it into `connect`, the nesting and complexity of that function would get a bit out of hand IMO. However, I can make the comment and names more explicit (e.g. `default` and `custom` instead of `left` and `right`)